core/src/main/java/uk/gov/justice/services/core/dispatcher/Dispatcher.java

```diff
@@ -25,7 +38,28 @@ public Dispatcher(final HandlerRegistry handlerRegistry) {
      * @return the envelope returned by the handler method
      */
     public JsonEnvelope dispatch(final JsonEnvelope envelope) {
-        return (JsonEnvelope) handlerRegistry.get(envelope.metadata().name()).execute(envelope);
+
+        final HandlerMethod handlerMethod = handlerRegistry.get(envelope.metadata().name());
+        final Class<?> envelopeGenericType = handlerMethod.getEnvelopeGenericType();
+
+        if(envelopeGenericType == JsonEnvelope.class) {
+            return (JsonEnvelope) handlerMethod.execute(envelope);
+        }
+
+        final JsonObject jsonObject = envelope.payloadAsJsonObject();
+        final Object readValue;
+
+        try {
+            readValue = objectMapper.readValue(jsonObject.toString(), envelopeGenericType);
+        } catch (IOException e) {
+            throw new HandlerExecutionException(
+                    format("Error while invoking handler method %s with parameter %s",
+                            handlerMethod, envelope), e.getCause());
+        }
+
+        final Envelope<Object> pojoEnvelope = (Envelope.envelopeFrom(envelope.metadata(), readValue));
+
+        return (JsonEnvelope) handlerMethod.execute(pojoEnvelope);
```

Comment:
We need to handle the fact that the return type of the handler methos might not be a JsonEnvelope